### PR TITLE
[TS migration] Migrate 'GrowlNotification' component to TypeScript

### PR DESCRIPTION
**Pull request by cdOut**
### Details

Migrated GrowlNotification and all the related components to TypeScript.

### Fixed Issues

$ https://github.com/Expensify/App/issues/25127

### Tests

1. Trigger the GrowlNotification (Since it's very difficult to trigger naturally in the app [depends on onfido backend errors or enabling the wallet which is available for US users only], you can add the snippet of code `Growl.show('Programatically created notification', CONST.GROWL.SUCCESS, 3000);` into useEffect of any rendered component. I added it to the WalletPage component)
2. Tap/Click on displayed notification
3. Verify that notification has disappeared after tapping/clicking

- [x] Verify that no errors appear in the JS console

### QA Steps

1. Make sure you are using an account with enabled wallet or enable it on your account
2. Go into the wallet statement page and press a button to download a wallet statement
3. After pressing it a growl notification should appear in the top right corner of the screen
4. Tap/Click on displayed notification
5. Verify that notification has disappeared after tapping/clicking

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/88325488/0d864238-8674-4270-9e26-1bc45e61bd6e

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/88325488/33817663-180f-4d40-bd20-f7545aa4f098

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/88325488/b9f3c841-016c-4e70-a8d4-57e1952e43c0

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/88325488/ea02a26f-b1e1-4786-815e-392923faf4d8

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/88325488/842f1151-2ce5-4eb6-849e-516825b1e0a1

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/88325488/43558948-a053-482d-b433-b3fa94a453fe

</details>

### CODE REVIEW

File: src/components/GrowlNotification/index.tsx
Comment (blazejkustra) [inline src/components/GrowlNotification/index.tsx:21]: It doesn't look right, have you tested it?

File: src/components/GrowlNotification/GrowlNotificationContainer/index.tsx
Comment (blazejkustra) [inline src/components/GrowlNotification/GrowlNotificationContainer/index.tsx:22]: Use a `useWindowDimensions` hook instead of the HOC

File: src/components/GrowlNotification/types.ts
Comment (blazejkustra) [inline src/components/GrowlNotification/types.ts:1]: Use type from [Growl.ts](https://github.com/software-mansion-labs/expensify-app-fork/blob/2b955ccd6600591dda3c3b8d0a92fd04ddc554f3/src/libs/Growl.ts) instead

File: src/components/GrowlNotification/index.tsx
Comment (blazejkustra) [reply src/components/GrowlNotification/index.tsx:21]: Have a [read](https://react.dev/reference/react/forwardRef) @cdOut! It should look more like this:
```ts
function GrowlNotification(_: unknown, ref: ForwardedRef<GrowlRef>) {
```

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/components/GrowlNotification/types.ts
Comment (cdOut) [reply src/components/GrowlNotification/types.ts:1]: I moved it here due to this comment:

https://github.com/Expensify/App/issues/25127#issuecomment-1715359171

File: src/components/GrowlNotification/index.tsx
Comment (blazejkustra) [inline src/components/GrowlNotification/index.tsx:64]: Why was this changed? Make sure to test it throughly 😃

File: src/components/GrowlNotification/GrowlNotificationContainer/types.ts
Comment (blazejkustra) [inline src/components/GrowlNotification/GrowlNotificationContainer/types.ts:4]: Use `ChildrenProps` type instead

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/components/GrowlNotification/index.tsx
Comment (cdOut) [reply src/components/GrowlNotification/index.tsx:64]: This variable does not exist for Animated.spring config parameter, it didn't serve any purpose.

File: src/components/GrowlNotification/types.ts
Comment (cdOut) [reply src/components/GrowlNotification/types.ts]: We agreed with Błażej to keep it where it was without moving it here.

**Review by blazejkustra — APPROVED** [review]
LGTM!

File: src/components/GrowlNotification/index.tsx
Comment (kubabutkiewicz) [inline src/components/GrowlNotification/index.tsx:30]: extract it to seperate type

**Comment by cdOut** [comment]
cc: @kubabutkiewicz

**Review by kubabutkiewicz — APPROVED** [review]
LGTM

**Comment by cdOut** [comment]
Please refer to this conversation: https://expensify.slack.com/archives/C01GTK53T8Q/p1704193571187889

This PR is currently blocked and I'm waiting for Bank Refactor PR to be merged into main, after which we should look into removing the Growl from the application and swapping it for internally declared functionality. After the PR is merged I will work on finding non-growl ways of replacing the old functionality.

**Comment by blazejkustra** [comment]
Let's fix lint and make this PR as ready for review @cdOut

**Comment by cdOut** [comment]
@blazejkustra working on it right now, currently writing test steps and recording videos.

File: src/components/GrowlNotification/index.tsx
Comment (cubuspl42) [inline src/components/GrowlNotification/index.tsx:22]: Is this because `_`? I'm not completely sure if the linter isn't in the right here. Do we commonly use this convention?

File: src/components/GrowlNotification/index.tsx
Comment (cubuspl42) [inline src/components/GrowlNotification/index.tsx:76]: Why were the braces removed? Do we depend on the return value?

File: src/components/GrowlNotification/index.tsx
Comment (blazejkustra) [reply src/components/GrowlNotification/index.tsx:22]: Yes, naming convention in eslint is setup so we can't use arguments like `_`, `_props` etc. Here props aren't used so `_` name is fine with me.

Comment (cubuspl42) [reply src/components/GrowlNotification/index.tsx:22]: **Our** ESLint is configured so we can't use arguments starting with `_`. So, either we accepted the defaults or explicitly decided to use this.

While I agree with you on some level, and I know the `_` naming convention is general, it seems that Expensify explicitly opted out of using it.

We should either remove this linter config or change the name.

Comment (blazejkustra) [reply src/components/GrowlNotification/index.tsx:22]: > While I agree with you on some level, and I know the _ naming convention is general, it seems that Expensify explicitly opted out of using it.

I configured this ESlint rule, and forgot about arguments starting with `_` (I configured it when there were no TS files in the codebase)

> We should either remove this linter config or change the name.

We already use arguments starting with `_` in multiple places. How about adjusting the ESlint rule so it allows arguments starting with `_`?

File: src/components/GrowlNotification/index.tsx
Comment (cdOut) [reply src/components/GrowlNotification/index.tsx]: I was able to remove them here due to there only being one statement in the return of the arrow function, in both implementations the fling animation worked exactly the same.

File: src/components/GrowlNotification/index.tsx
Comment (cubuspl42) [reply src/components/GrowlNotification/index.tsx:22]: > How about adjusting the ESlint rule so it allows arguments starting with _?

Fine for me! I'm just ensuring that we don't fight with our own configuration for no good reason.

Comment (cdOut) [reply src/components/GrowlNotification/index.tsx]: @cubuspl42 @blazejkustra I've updated the rule to allow the usage of leading underscores and cleaned up the eslint-disables that were used as a workaround.

**Comment by cubuspl42** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos

<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/2590174/e0796acb-063c-4229-96b7-4acef763307a

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/2590174/ac1376db-a39b-46ef-b99e-7f78f0ea28b7

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/2590174/b7b93f08-06b1-4f39-9411-1eb43f507de2

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/2590174/ef4c0df4-dff9-44fb-b934-3438f0cc43fb

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/2590174/4bdc91cf-db7a-4eb1-8481-ae1241609149

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/2590174/ff9df58d-7309-44b0-9570-8bcca7255e3b

</details>

**Review by jakub-trzebiatowski — APPROVED** [review]
As a non-US person, I also had issues with Wallet configuration in the past. Maybeee we could try triggering an error in the sandbox, but I'm unsure whether it would work. I think that testing this programmatically will be good enough here.

**Review by jakub-trzebiatowski — APPROVED** [review]

File: src/components/GrowlNotification/index.tsx
Comment (hayata-suenaga) [inline src/components/GrowlNotification/index.tsx:22]:
```suggestion
function GrowlNotification(_: never, ref: ForwardedRef<GrowlRef>) {
```

File: src/components/GrowlNotification/index.tsx
Comment (hayata-suenaga) [inline src/components/GrowlNotification/index.tsx:26]:
```suggestion
    const [duration, setDuration] = useState<number>();
```

You don't have to explicitly specify `undefined` if you leave the initial value of the state (i.e. the argument to `useState`) empty.

File: src/components/GrowlNotification/index.tsx
Comment (hayata-suenaga) [inline src/components/GrowlNotification/index.tsx:31]: Can you add a comment on what the key of the `GrowlIconTypes` represents?

Comment (cdOut) [reply src/components/GrowlNotification/index.tsx:34]: I've added a comment to clarify that part of the code.

**Comment by cdOut** [comment]
@hayata-suenaga I've implemented your review suggestions, could you please take a look at this PR again?

File: src/components/GrowlNotification/index.tsx
Comment (cdOut) [reply src/components/GrowlNotification/index.tsx]: Not possible, prop type for JSX components cannot be never as shown in the following screenshot. I'm leaving it as unknown.

Screenshot 2024-03-12 at 16 29 24

**Comment by cdOut** [comment]
@hayata-suenaga I've implemented your review suggestions, could you please check the PR again?

**Review by hayata-suenaga — APPROVED** [review]
🟢

**Comment by hayata-suenaga** [comment]
All tests were passing. don't know why Melvin thinks otherwise 🤔 

Screenshot 2024-03-12 at 1 34 45 PM

**Pull request merged**
